_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+  sentinel feature. However, being a stateful server, there is a high maintenance cost to keeping it up and running, in
+  addition to licensing cost for using the service.
+
+In this guide, we will use the ECS Deploy Runner Stack as the deploy server. Note that although we will not explicitly
+cover it, the design is compatible with using Terraform Enterprise as the deploy server.
+
+
+[[limit_triggers_for_deploy_server]]
+=== Limit triggers for deploy server
+
+The deploy server should only expose a limited set of options for triggering deployments. That is, it should not allow
+arbitrary deployments on arbitrary code. For example, the default configuration of Atlantis allows webhooks from any
+repository. This means that any public repo can cause your Atlantis server to run `terraform plan` and `terraform apply`
+on custom code you do not control using the permissions granted to that server. Instead, you will want to configure it
+so that only certain repositories, branches, and users can trigger the workflow.
+
+The Gruntwork ECS Deploy Runner stack mitigates this concern by only allowing triggers from a Lambda function that
+exposes a limited set of actions against the deploy runner task. The lambda function:
+
+- Requires a single repository to trigger deployments by default.
+- Can be configured to limit deployments to specific branches.
+- Requires explicit IAM permissions to trigger.
+
+You can find similar mechanisms for limiting deployments in the various deploy server options.
+
+
+[[use_a_vpc_to_lock_down_infrastructure_deployer]]
+=== Use a VPC to lock down deploy server
+
+Run your infrastructure deployment workloads in a https://aws.amazon.com/vpc/[Virtual Private Cloud (VPC)] to isolate
+the workloads in a restricted network topology (see link:/guides/networking/how-to-deploy-production-grade-vpc-aws[How
+to deploy a production-grade VPC on AWS] for more information on VPCs). Configure to run all workloads in private
+subnets that are not publicly accessible. Make sure to block all inbound internet access and consider blocking all
+outbound access except for the minimum required (e.g, allow access to AWS APIs).
+
+
+[[use_minimal_iam_permissions_for_a_deployment]]
+=== Use minimal IAM permissions for a deployment
+
+Avoid having a single system with admin permissions for running a deployment. Instead, deploy specialized versions of
+the deployment platforms with varying permissions for handling specific workflows. By separating out the concerns for
+each pipeline, you can reduce the blast radius of the damage that can be done with each set of credentials. At a minimum,
+you should have two versions of the infrastructure deployment system: one for deploying the application code, and which
+only has the minimal permissions necessary for deploying that application; and one for deploying infrastructure code,
+which has full access to the environments.
```

Comment:
```suggestion
which has more access to the environments.
```

Even for the infra deployments, it would be ideal if it could have reduced permissions. In fact, to be compliant with CIS, you couldn't have a deployment system with full access.